enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,232 @@
+
+#### Kubelet
+
+Openshift manages `kubelet` configuration with `machine-config-operator`. The actual `cloud-config` arguments are passed externally from the `Infrastructure` resource on https://github.com/openshift/machine-config-operator/blob/2a51e49b0835bbd3ac60baa685299e86777b064f/pkg/controller/template/render.go#L310-L357, where the support for `external` cloud provider will be added. For each platform with `external` configuration, the `cloud-config` flag will not be set.
+
+#### Kube-controller-manager
+
+This configuration is managed by the `cluster-kube-controller-manager-operator` code and will need to integrate the `library-go` changes, described above in the operator [configuration](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85).
+
+#### Kube-api-server
+
+No changes for the current configuration are needed
+
+### Operator resource management strategy
+
+Every cloud manages it's configuration differently, but share a common overall structure.
+
+1. `cloud-controller-manager` and related resources are deployed under one namespace.
+2. Each pod (`ccm`, `cloud-node-manager`, etc.) supports deployment with a `DaemonSet`.
+3. The pods are expected to run on `Nodes` marked as a control-plane with annotation:
+```yaml
+    node-role.kubernetes.io/master: ""
+```
+4. The pods are sharing common command line arguments with the `kcm` pods, such as `--cloud-provider` and `--cloud-config`.
+
+Proposed operator will not diverge from the upstream resource structuring:
+
+- Static infrastructure resources creation, such as namespaces or service accounts will be delegated to the [CVO](https://github.com/openshift/cluster-version-operator).
+- Operator will populate, create and manage the cloud-specific set of `DaemonSets`, based on the selected `platform` in the `Infrastructure` resource.
+- The `--cloud-provider` and the `--cloud-config` arguments will be populated using existing `library-go` based implementation from the [cluster-kube-controller-manager-operator](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85).
+
+#### Examples for the cloud-controller-manager configurations
+
+##### AWS
+
+Main repository: https://github.com/kubernetes/cloud-provider-aws/
+Sample manifests: https://github.com/kubernetes/cloud-provider-aws/tree/master/manifests
+
+To try external cloud-controller-manager, simply run:
```

Comment:
This mentions trying this, is that a phrasing from upstream or do they expect that these resources are production ready and this is how they'd recommend deploying in production?

Reply:
As Stefan mentioned, we have a completely different set of goals for Openshift, so their decisions are not applying to us. I removed this section for that reason, and we should probably settle for the combination of static pods during bootstrap and a Deployment at runtime phase.